### PR TITLE
Add QuadrilinearFieldInterpolator

### PR DESCRIPTION
**Pull request by brittonsmith**
## PR Summary

This adds a `QuadrilinearFieldInterpolator` class to do 4d linear interpolation. This work done primarily by Evan Jones, a masters student working with me over the last year. Evan has moved on, so I'm pushing this forward to get it into the codebase. This functionality is required for some new features added to a yt extension package, but this still seems like the most appropriate place for the base code. I've added a test analogous to the ones for the other field interpolators. Note, I specifically adopt the field ordering of "xyzw" instead of "wxyz" for consistency with the function docs and naming convention in the other tests.

## PR Checklist

- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by brittonsmith** [comment]
pre-commit.ci autofix

File: yt/utilities/linear_interpolators.py [hunk 1]
```diff
@@ -167,10 +167,10 @@ def __init__(self, table, boundaries, field_names, truncate=False):
 
         ad = ds.all_data()
         table_data = np.random.random((64, 64, 64))
-        interp = BilinearFieldInterpolator(table_data,
-                                           (0.0, 1.0, 0.0, 1.0, 0.0, 1.0),
-                                           ["x", "y", "z"],
-                                           truncate=True)
+        interp = TrilinearFieldInterpolator(table_data,
```
Comment (cphyc) [inline yt/utilities/linear_interpolators.py:170]: Nice catch!

File: yt/utilities/lib/interpolators.pyx [hunk 2]
```diff
@@ -110,6 +110,67 @@ def TrilinearlyInterpolate(np.ndarray[np.float64_t, ndim=3] table,
...
+        z_i = z_is[i]
+        w_i = w_is[i]
+        x = x_vals[i]
+        y = y_vals[i]
+        z = z_vals[i]
+        w = w_vals[i]
+        dx_inv = 1.0 / (x_bins[x_i+1] - x_bins[x_i])
+        dy_inv = 1.0 / (y_bins[y_i+1] - y_bins[y_i])
+        dz_inv = 1.0 / (z_bins[z_i+1] - z_bins[z_i])
+        dw_inv = 1.0 / (w_bins[w_i+1] - w_bins[w_i])
```
Comment (cphyc) [inline yt/utilities/lib/interpolators.pyx:148]: A note on performance: it might be less wasteful to store these intermediate values in a temporary array to prevent them from being recomputed over and over.

File: yt/fields/derived_field.py [hunk 3]
```diff
@@ -389,6 +389,16 @@ def _ion_to_label(self):
...
+            "20": "XXI",
+            "21": "XXII",
+            "22": "XXIII",
+            "23": "XXIV",
+            "24": "XXV",
+            "25": "XXVI",
+            "26": "XXVII",
+            "27": "XXVIII",
+            "28": "XXIX",
+            "29": "XXX",
```
Comment (cphyc) [inline yt/fields/derived_field.py:401]: Why is this change necessary?

**Review by cphyc — APPROVED** [review]

**Review by jzuhone — APPROVED** [review]
Other than agreeing with the comment about performance, LGTM

File: yt/fields/derived_field.py (hunk 3 above)
Comment (matthewturk) [reply yt/fields/derived_field.py:401]: If I were to guess, it's that the ionization states are used in the same place as this new field interpolator

File: yt/utilities/lib/interpolators.pyx (hunk 2 above)
Comment (matthewturk) [reply yt/utilities/lib/interpolators.pyx:148]: Hm, that's the case if they're equally spaced bins, but do we mandate that?  They really might vary on the iteration.

**Review by matthewturk — APPROVED** [review]

File: yt/fields/derived_field.py (hunk 3 above)
Comment (brittonsmith) [reply yt/fields/derived_field.py:401]: Yeah, that's right. Sometimes, you just need more ions.
